### PR TITLE
Make default hosted model optional for k8s controller.

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

Now we do not create the `default hosted model` for `k8s controller`, 
but the user can still request a preferred hosted model created 
during bootstrapping time by specifying option `--default-model`.

## QA steps

- add k8s cloud

```bash
$ microk8s.config | juju add-k8s <k8s-cloud>
```

- juju bootstrap without hosted model created;

```bash
$ juju bootstrap <k8s-cloud> <controller-name> --debug
```

- juju bootstrap with hosted model created (the "default" model name will be valid for k8s controller);

```bash
$ juju bootstrap <k8s-cloud> <controller-name> --debug --default-model <my-default-hosted-model>
```

## Documentation changes

Added instructions in Bootstrapping Infof messages;

## Bug reference

None

### CODE REVIEW

File: agent/agentbootstrap/bootstrap.go [hunk 1]
```diff
@@ -191,6 +191,10 @@ func ensureHostedModel(
 	adminUser names.UserTag,
 	cloudCredentialTag names.CloudCredentialTag,
 ) error {
+	if args.HostedModelConfig == nil || len(args.HostedModelConfig) == 0 {
```
Comment (wallyworld) [inline agent/agentbootstrap/bootstrap.go]: Just need len()
No need for extra nil check

Comment (wallyworld) [reply agent/agentbootstrap/bootstrap.go]: Also don't need the IsCAAS arg any more

File: cloudconfig/podcfg/podcfg.go [hunk 2]
```diff
@@ -408,7 +405,9 @@ func NewBootstrapControllerPodConfig(config controller.Config, controllerName, s
 	}
 	pcfg.Jobs = []multiwatcher.MachineJob{
 		multiwatcher.JobManageModel,
-		multiwatcher.JobHostUnits,
+	}
+	if pcfg.Bootstrap.HostedModelConfig != nil && len(pcfg.Bootstrap.HostedModelConfig) > 0 {
```
Comment (wallyworld) [inline cloudconfig/podcfg/podcfg.go]: Don't need this check - the controller model can always host units

File: cmd/juju/commands/bootstrap.go [hunk 3]
```diff
@@ -564,45 +568,69 @@ func (c *bootstrapCommand) Run(ctx *cmd.Context) (resultErr error) {
...
-		return errors.Trace(err)
-	}
+	getHostedModel := func() (*jujuclient.ModelDetails, error) {
+		if isCAASController && c.hostedModelName == defaultHostedModelName {
+			// k8s controller does NOT have "default" hosted model
+			// if the user didn't specify a preferred hosted model name.
+			initialHostedModelInstructionMsg = `
+Now you can run 
+	juju add-model <model-name>
+to create a new model to deploy k8s workloads
```
Comment (wallyworld) [inline cmd/juju/commands/bootstrap.go]: End with "."

File: cmd/juju/commands/bootstrap.go [hunk 4]
```diff
@@ -564,45 +568,69 @@ func (c *bootstrapCommand) Run(ctx *cmd.Context) (resultErr error) {
 		return errors.Trace(err)
 	}
 
-	hostedModelUUID, err := utils.NewUUID()
-	if err != nil {
-		return errors.Trace(err)
-	}
+	getHostedModel := func() (*jujuclient.ModelDetails, error) {
```
Comment (wallyworld) [inline cmd/juju/commands/bootstrap.go]: This is large enough that it should be a non-inline func

File: agent/agentbootstrap/bootstrap.go (hunk 1 above)
Comment (ycliuhw) [reply agent/agentbootstrap/bootstrap.go]: yeah,
we still need `isCAAS` to get environ

File: cloudconfig/podcfg/podcfg.go (hunk 2 above)
Comment (ycliuhw) [reply cloudconfig/podcfg/podcfg.go]: fixed

File: cmd/juju/commands/bootstrap.go (hunk 3 above)
Comment (ycliuhw) [reply cmd/juju/commands/bootstrap.go]: fixed

File: cmd/juju/commands/bootstrap.go (hunk 4 above)
Comment (ycliuhw) [reply cmd/juju/commands/bootstrap.go]: agreed, moved to a separate method

File: cmd/juju/common/controller_test.go [hunk 5]
```diff
@@ -116,7 +116,7 @@ func (s *controllerSuite) TestWaitForAgentAPIReadyWaitsForSpaceDiscovery(c *gc.C
 	s.mockBlockClient.discoveringSpacesError = 2
 
 	runInCommand(c, func(ctx *cmd.Context, base *modelcmd.ModelCommandBase) {
-		err := WaitForAgentInitialisation(ctx, base, "controller", "default")
+		err := WaitForAgentInitialisation(ctx, base, false, "controller", "default")
```
Comment (ycliuhw) [inline cmd/juju/common/controller_test.go]: the only difference for `caas` and `iaas` is the Infof messages, so no test for caas case.

**Comment by ycliuhw** [comment]
[solve k8s controller default hosted model conflict - k8s has default namespace already.](https://trello.com/c/QNYTbjRm/673-solve-k8s-controller-default-hosted-model-conflict-k8s-has-default-namespace-already)

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**